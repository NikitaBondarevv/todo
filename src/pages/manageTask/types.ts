import { IData } from 'interfaces/IData'

export type TTaskProps = {
  onSubmit: (data: IData) => Promise<void>
  text?: string
}

export type TTarget = {
  target: {
    value: string
  }
}
