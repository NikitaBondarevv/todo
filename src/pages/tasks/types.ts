export interface TData {
  day: string
  id: string
  title: string
  done: boolean
}

export type TTaskProps = {
  data: TData
  getTasks: () => void
}

export type TTaskListProps = {
  tasks: TData[]
  getTasks: () => void
  activeTabIndex: number
}
