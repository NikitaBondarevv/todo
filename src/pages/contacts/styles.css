.contacts {
  border: 1px solid #fff;
  box-shadow: 0 0 5px 1px rgb(0 0 0 / 50%);
  padding: 39px 0 0;
  text-align: center;
  background: rgb(7 21 147 / 50%);
  width: 835px;
  color: #fff;
  font: 35px/40px "MyriadProRegular";
  margin: 0 auto 100px;
  letter-spacing: 1px;
  height: 487px;
}

.email {
  background: none;
  color: #fff;
  font-weight: bold;
  text-decoration: none;
  padding: 0 0 0 5px;
}

.email:hover {
  text-decoration: underline;
}
