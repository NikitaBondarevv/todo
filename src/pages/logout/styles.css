.logout {
  border: 1px solid #fff;
  box-shadow: 0 0 5px 1px rgb(0 0 0 / 50%);
  padding: 39px 29px 104px;
  background: rgb(7 21 147 / 50%);
  width: 777px;
  height: 383px;
  color: #fff;
  font: 24px/40px "MyriadProRegular";
  margin: 0 auto 100px;
  letter-spacing: 1px;
}

.error {
  color: red;
  font-size: 25px;
  margin: 0 0 0 5px;
}
