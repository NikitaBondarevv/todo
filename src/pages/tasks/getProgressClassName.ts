import styles from './styles.css'

export const getProgressClassName = (done: boolean | undefined) => {
  if (done) return styles.done

  if (done === undefined) return `${styles.progress} ${styles.updateTask}`

  return `${styles.waiting} ${styles.updateTask}`
}
