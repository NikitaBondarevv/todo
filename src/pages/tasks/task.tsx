import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'

import { updateTask, deleteTask } from 'contracts/tasks'
import { getProgressClassName } from './getProgressClassName'
import { TTaskProps } from './types'
import styles from './styles.css'

export const Task = ({ data, getTasks }: TTaskProps) => {  
  const { done, title } = data

  const completeTask = async () => {
    await updateTask({ ...data, done: true })
    getTasks()
  }
  
  const setInProgress = async () => {
    await updateTask({ ...data, done: undefined })
    getTasks()
  }

  const removeTask = async () => {
    await deleteTask(data)
    getTasks()
  }

  return (
    <li className={getProgressClassName(done)}>
      &nbsp;
      <Link to={`/task/${data.day}/${data.id}`}>{title}</Link>
      {
        !done && (
          <div className={styles.buttons}>
            <button onClick={completeTask} type="button" />
            {
              done !== undefined && (
                  <button onClick={setInProgress} className={styles.inProgress} type="button" />
                )
            }
            <button onClick={removeTask} className={styles.delete} type="button" />
          </div>
        )
      }
    </li>
  )
}

Task.defaultProps = {
  data: {},
}

Task.propTypes = {
  data: PropTypes.object,
  getTasks: PropTypes.func
}
