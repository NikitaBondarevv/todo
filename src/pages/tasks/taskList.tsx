import { useState, useEffect } from 'react'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'

import { Task } from './task'
import { TTaskListProps } from './types'
import styles from './styles.css'

export const TaskList = ({ tasks, getTasks, activeTabIndex }: TTaskListProps) => {
  const [todos, setTodos] = useState(tasks)

  useEffect(() => {
    setTodos(tasks)
  }, [tasks])

  return (
    <div>
      <ol className={styles.tasksList}>
        {
          todos.map((task, index) => <Task key={index} data={task} getTasks={getTasks} />)
        }
      </ol>
      <Link to={`/task/${activeTabIndex}`} className={styles.addTask}>Add new task</Link>
    </div>
  )
}

TaskList.defaultProps = {
  tasks: [],
  activeTabIndex: 0,
  getTasks: () => {}
}

TaskList.propTypes = {
  tasks: PropTypes.array,
  getTasks: PropTypes.func,
  activeTabIndex: PropTypes.number
}
