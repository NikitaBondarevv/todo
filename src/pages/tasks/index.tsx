import { useState, useEffect } from 'react'

import { daysOfTheWeek } from 'helpers/constans'
import { getTasks } from 'contracts/getTasks'
import { Preloader } from 'components/preloader'
import { TaskList } from './taskList'
import { Tabs } from 'components/tabs'
import { Tab } from 'components/tabs/tab'

export const Tasks = () => {
  const [days, setDays] = useState([])
  const [isLoading, setIsloading] = useState(false)

  const getDailyTasks = async () => {
    setDays(await getTasks())
    setIsloading(false)
  }

  useEffect(() => {
    setIsloading(true)
    getDailyTasks()
  }, [])

  const getCurrentDay = () => {
    const dayNow = new Date().getDay()

    return dayNow ? dayNow - 1 : 6
  }

  return (
    isLoading ? <Preloader size={100} />
      : <Tabs selectedIndex={getCurrentDay()}>
        {
          days.map((day, index) => (
            <Tab key={index} title={daysOfTheWeek[index]}>
              <TaskList tasks={day} getTasks={getDailyTasks} activeTabIndex={index} />
            </Tab>
          ))
        }
      </Tabs>
  )
}
