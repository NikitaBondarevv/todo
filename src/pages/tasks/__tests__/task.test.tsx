import { act, fireEvent, render, screen } from '@testing-library/react'
import { BrowserRouter } from 'react-router-dom'

import { task } from '__mocks__/entities/task.mock'
import { Task } from '../task'
import { TTaskProps } from '../types'
import { deleteTask, updateTask } from 'contracts/tasks'
import styles from '../styles.css'

jest.mock('contracts/tasks', () => ({
  updateTask: jest.fn(),
  deleteTask: jest.fn()
}))

describe('<Task />', () => {
  const defaultProps: TTaskProps = {
    data: task,
    getTasks: jest.fn()
  }
  const renderWrapper = (props?: Partial<TTaskProps>, container?: HTMLElement) => render(
    <BrowserRouter>
      <Task {...defaultProps} {...props} />
    </BrowserRouter>,
    { container }
  )

  beforeEach(() => {
    jest.resetAllMocks()
  })

  test('should match snapshot', () => {
    const { asFragment } = renderWrapper()

    expect(asFragment()).toMatchSnapshot()
  })

  test('should call updateTask() when click completeTask button', async () => {
    renderWrapper()
    const buttons = screen.getAllByRole('button')

    await act(() => fireEvent.click(buttons.at(0)!))

    expect(updateTask).toHaveBeenCalledWith({ ...task, done: true })
    expect(defaultProps.getTasks).toHaveBeenCalled()
  })

  test('should call updateTask() when click setInProgress button', async () => {
    const container = document.createElement('div')
    renderWrapper({ data: { ...task, done: false } }, container)
    const button = container.querySelector(`.${styles.inProgress}`)

    await act(() => fireEvent.click(button!))

    expect(updateTask).toHaveBeenCalledWith({ ...task, done: undefined })
    expect(defaultProps.getTasks).toHaveBeenCalled()
  })

  test('should call deleteTask() when click removeTask button', async () => {
    renderWrapper()
    const buttons = screen.getAllByRole('button')

    await act(() => fireEvent.click(buttons.at(2)!))

    expect(deleteTask).toHaveBeenCalledWith(task)
    expect(defaultProps.getTasks).toHaveBeenCalled()
  })
})
