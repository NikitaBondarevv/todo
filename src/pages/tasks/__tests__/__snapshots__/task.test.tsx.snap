// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Task /> should match snapshot 1`] = `
<DocumentFragment>
  <li
    class="waiting updateTask"
  >
    Â 
    <a
      href="/task/0/0"
    >
      test
    </a>
    <div
      class="buttons"
    >
      <button
        type="button"
      />
      <button
        class="inProgress"
        type="button"
      />
      <button
        class="delete"
        type="button"
      />
    </div>
  </li>
</DocumentFragment>
`;
